core/reactor.c

```diff
@@ -315,24 +272,18 @@ int next(void) {
 
     LF_PRINT_LOG("Next event (elapsed) time is " PRINTF_TIME ".", next_tag.time - start_time);
     // Wait until physical time >= event.time.
-    // The wait_until function will advance current_tag.time.
     if (wait_until(next_tag.time) != 0) {
         LF_PRINT_DEBUG("***** wait_until was interrupted.");
-        // Sleep was interrupted.
-        // FIXME: It is unclear what would cause this to occur in this unthreaded
-        // runtime since lf_schedule() is not thread safe here and should not
-        // be called asynchronously. Perhaps in some runtime such as for a
-        // PRET machine this will be supported, so here we handle this as
-        // if an asynchronous call to schedule has occurred. In that case,
-        // we should return 1 to let the runtime loop around to see what
-        // is on the event queue.
+        // Sleep was interrupted. This could happen when a physical action
+        // gets scheduled from an interrupt service routine.
         return 1;
     }
 
     // At this point, finally, we have an event to process.
     // Advance current time to match that of the first event on the queue.
+    lf_critical_section_enter();
```

Comment:
Seems like there is a race condition here. We could get an interrupt between when wait_until returns and here, which means that the next_tag may be obsolete now? Actually, maybe it's OK because if wait_until was not interrupted, then the physical clock will have advanced past next_tag, so even if an interrupt occurred, any events it would have created would have a bigger tag?

Reply:
I think your concern about the possibility of a race condition has been addressed in a7d20295f7b74a5e4f67beb80494419c38332a30.